Error in Adafruit HTML crashes application

```
/node_modules/cheerio/lib/parse.js:55
stock_fetch_1  |     var oldParent = node.parent || node.root,
stock_fetch_1  |                         ^
stock_fetch_1  | 
stock_fetch_1  | TypeError: Cannot read property 'parent' of undefined
stock_fetch_1  |     at Function.exports.update (/node_modules/cheerio/lib/parse.js:55:25)
stock_fetch_1  |     at module.exports (/node_modules/cheerio/lib/parse.js:17:11)
stock_fetch_1  |     at Function.exports.load (/node_modules/cheerio/lib/static.js:19:14)
stock_fetch_1  |     at Request._callback (/fetch/adafruit.js:12:36)
stock_fetch_1  |     at self.callback (/node_modules/request/request.js:200:22)
stock_fetch_1  |     at emitOne (events.js:77:13)
stock_fetch_1  |     at Request.emit (events.js:169:7)
stock_fetch_1  |     at Request.onRequestError (/node_modules/request/request.js:831:8)
stock_fetch_1  |     at emitOne (events.js:77:13)
stock_fetch_1  |     at ClientRequest.emit (events.js:169:7)
stock_fetch_1  | 
stock_fetch_1  | npm ERR! Linux 4.4.0-21-generic
stock_fetch_1  | npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "start"
stock_fetch_1  | npm ERR! node v4.4.4
stock_fetch_1  | npm ERR! npm  v2.15.5
stock_fetch_1  | npm ERR! code ELIFECYCLE
stock_fetch_1  | npm ERR! stock_fetch@1.0.0 start: `node index.js`
stock_fetch_1  | npm ERR! Exit status 1
stock_fetch_1  | npm ERR! 
stock_fetch_1  | npm ERR! Failed at the stock_fetch@1.0.0 start script 'node index.js'.
stock_fetch_1  | npm ERR! This is most likely a problem with the stock_fetch package,
stock_fetch_1  | npm ERR! not with npm itself.
stock_fetch_1  | npm ERR! Tell the author that this fails on your system:
stock_fetch_1  | npm ERR!     node index.js
stock_fetch_1  | npm ERR! You can get information on how to open an issue for this project with:
stock_fetch_1  | npm ERR!     npm bugs stock_fetch
stock_fetch_1  | npm ERR! Or if that isn't available, you can get their info via:
stock_fetch_1  | npm ERR! 
stock_fetch_1  | npm ERR!     npm owner ls stock_fetch
stock_fetch_1  | npm ERR! There is likely additional logging output above.
stock_fetch_1  | 
stock_fetch_1  | npm ERR! Please include the following file with any support request:
stock_fetch_1  | npm ERR!     /npm-debug.log
```

Suggest try/catch or defensive coding.

--- a/stock_fetch/fetch/adafruit.js
+++ b/stock_fetch/fetch/adafruit.js
@@ -7,18 +7,19 @@ class Adafruit {
     this.url = url;
   }
 
-  refresh(done) {
-    this.modules.request.get(this.url, (err,res,body)=> {
-      let $ = this.modules.cheerio.load(body);
-      let rows = $(".product-listing a");
+  _scrape($) {
+    let rows = $(".product-listing a");
 
-      let zeroProducts = this.zeroProducts;
-      let stock = false;
+    let zeroProducts = this.zeroProducts;
+    let stock = false;
+    if(rows) {
 
       rows.each((i, r)=> {
-        if(zeroProducts.indexOf(r.attribs["data-pid"]) > -1) {
+        if(r && r.attribs && r.attribs["data-pid"] && 
+          zeroProducts.indexOf(r.attribs["data-pid"]) > -1) {
+
           let isImageLink = false;
-          r.children.forEach((ch)=>{
+          r.children.forEach((ch) => {
             if(ch.type != "text") {
               isImageLink = true;
             }
@@ -27,13 +28,33 @@ class Adafruit {
           if(!isImageLink) {
             // console.log("PI Zero found - " + r.attribs['data-name'])
             // console.log(r.attribs["data-pid"]);
-            var available = $(r.parent.parent.parent).html().indexOf("out-of-stock") == -1;
-            if(available) {
-              stock = true;
+            if(r.parent && r.parent.parent && r.parent.parent.parent) {
+              var available = $(r.parent.parent.parent).html().indexOf("out-of-stock") == -1;
+              if(available) {
+                stock = true;
+              }
             }
           }
         }
       });
+    }
+
+    return stock;
+  }
+
+  refresh(done) {
+    this.modules.request.get(this.url, (err,res,body)=> {
+      let $ = null;
+      let stock = false;
+      try { 
+        $ = this.modules.cheerio.load(body);
+        if($) {
+          stock = this._scrape($);
+        }
+      }
+      catch (e) {
+        console.error(e);
+      }
 
       done(err, {stock:stock});
     });

--- a/stock_fetch/test/adafruitTest.js
+++ b/stock_fetch/test/adafruitTest.js
@@ -12,6 +12,79 @@ describe("adafruitTests", () => {
     sandbox = sinon.sandbox.create();
     request = { get: sinon.stub() };
   });
+  
+  it("has badly formed HTML -> no rows", (done) => {
+      var fakecheerio = {
+        load : sinon.stub()
+      };
+      var fakeDoc = (docKey) => {
+        if(docKey== ".product-listing a") {
+          return null;
+        }
+      };
+
+      fakecheerio.load.returns(fakeDoc);
+      request.get.yields(null, {}, "<html></html>");
+      let finder = new Adafruit({request: request,cheerio: fakecheerio}, ["2816", "2817", "2885"],  "http://webpage/");
+
+      finder.refresh((err, stock) => {
+        expect(err).not.to.exist;
+        expect(stock).to.exist;
+        expect(stock.stock).to.equal(false);
+        expect(stock.totalAmount).not.to.exist;
+        done();
+      });
+  });
+  
+
+  it("fails to parse HTML", (done) => {
+
+      var fakecheerio = {
+        load : sinon.stub()
+      };
+      var eachStub = sinon.stub();
+      eachStub.yields(0, null);
+      
+      fakecheerio.load.throws("cannot parse this document");
+      request.get.yields(null, {}, "<html></html>");
+      let finder = new Adafruit({request: request,cheerio: fakecheerio}, ["2816", "2817", "2885"],  "http://webpage/");
+
+      finder.refresh((err, stock) => {
+        expect(err).not.to.exist;
+        expect(stock).to.exist;
+        expect(stock.stock).to.equal(false);
+        expect(stock.totalAmount).not.to.exist;
+        done();
+      });
+  });
+
+  it("has badly formed HTML -> has no attribs in row", (done) => {
+      var fakecheerio = {
+        load : sinon.stub()
+      };
+      var eachStub = sinon.stub();
+      eachStub.yields(0, null);
+      
+      var fakeDoc = (docKey) => {
+        if(docKey== ".product-listing a") {
+          return {
+            each : eachStub
+          };
+        }
+      };
+
+      fakecheerio.load.returns(fakeDoc);
+      request.get.yields(null, {}, "<html></html>");
+      let finder = new Adafruit({request: request,cheerio: fakecheerio}, ["2816", "2817", "2885"],  "http://webpage/");
+
+      finder.refresh((err, stock) => {
+        expect(err).not.to.exist;
+        expect(stock).to.exist;
+        expect(stock.stock).to.equal(false);
+        expect(stock.totalAmount).not.to.exist;
+        done();
+      });
+  });
 
   it("finds product stock", (done) => {
 

